fish-rust/src/builtins/cd.rs

```diff
@@ -0,0 +1,181 @@
+            }
+            None => {
+                streams
+                    .err
+                    .append(wgettext_fmt!("%ls: Could not find home directory\n", cmd));
+                return STATUS_CMD_ERROR;
+            }
+        }
+    };
+
+    // Stop `cd ""` from crashing
+    if dir_in.is_empty() {
+        streams.err.append(wgettext_fmt!(
+            "%ls: Empty directory '%ls' does not exist\n",
+            cmd,
+            dir_in
+        ));
+        // XXX why is this not .libdata_pod().is_interactive, does it matter?
+        if !parser.is_interactive() {
+            streams.err.append(parser.pin().current_line().from_ffi());
+        };
+        return STATUS_CMD_ERROR;
+    }
+
+    let pwd = vars.get_pwd_slash();
+
+    let dirs = path_apply_cdpath(dir_in, &pwd, vars.as_ref());
+    if dirs.is_empty() {
+        streams.err.append(wgettext_fmt!(
+            "%ls: The directory '%ls' does not exist\n",
+            cmd,
+            dir_in
+        ));
+
+        if !parser.is_interactive() {
+            streams.err.append(parser.pin().current_line().from_ffi());
+        }
+
+        return STATUS_CMD_ERROR;
+    }
+
+    let mut best_errno = 0;
+    let mut broken_symlink = WString::new();
+    let mut broken_symlink_target = WString::new();
+
+    for dir in dirs {
+        let norm_dir = normalize_path(&dir, true);
+
+        errno::set_errno(Errno(0));
+
+        // We need to keep around the fd for this directory, in the parser.
+        let dir_fd = AutoCloseFd::new(wopen_cloexec(&norm_dir, O_RDONLY, 0));
+
+        // Rust doesn't have library support for fchdir; set_current_dir could create a
+        // race condition (but there's a race below anyway - perhaps not important)
+        if !(dir_fd.is_valid() && unsafe { fchdir(dir_fd.fd()) } == 0) {
+            // Some errors we skip and only report if nothing worked.
+            // ENOENT in particular is very low priority
+            // - if in another directory there was a *file* by the correct name
+            // we prefer *that* error because it's more specific
+            if errno::errno().0 == ENOENT {
+                let tmp = wreadlink(&norm_dir);
+                // clippy doesn't like this is_some/unwrap pair, but using if let is harder to read IMO
+                #[allow(clippy::unnecessary_unwrap)]
+                if broken_symlink.is_empty() && tmp.is_some() {
+                    broken_symlink = norm_dir;
+                    broken_symlink_target = tmp.unwrap();
+                } else if best_errno == 0 {
+                    best_errno = errno::errno().0;
+                }
+                continue;
+            } else if errno::errno().0 == ENOTDIR {
+                best_errno = errno::errno().0;
+                continue;
+            }
+            best_errno = errno::errno().0;
+            break;
+        }
+
+        // Port note: sending the AutocloseFd across the FFI interface requires additional work
```

Comment:
I would prefer to fix this now. I made a PR-against-your-PR if you want to take it. [Here's the commit.](https://github.com/zanchey/fish-shell/pull/3/commits/38c72f0e59e2feab26e7481ec0db073b71a4e025)